dll/win32/shell32/CFolderItems.cpp

```diff
@@ -240,40 +252,51 @@ HRESULT STDMETHODCALLTYPE CFolderItems::Item(VARIANT index, FolderItem **ppid)
     if (!m_EnumIDList)
         return E_FAIL;
 
-    if (V_VT(&index) != VT_I4 && V_VT(&index) != VT_UI4)
-        return E_INVALIDARG;
+    if (V_VT(&index) == VT_I2)
+        VariantChangeType(&index, &index, 0, VT_I4);
 
-    ULONG count = V_UI4(&index);
+    if (V_VT(&index) == VT_I4)
+    {
+        ULONG count = V_UI4(&index);
 
-    HRESULT hr = m_EnumIDList->Reset();
-    if (FAILED_UNEXPECTEDLY(hr))
-        return hr;
+        HRESULT hr = m_EnumIDList->Reset();
+        if (FAILED_UNEXPECTEDLY(hr))
+            return hr;
 
-    hr = m_EnumIDList->Skip(count);
+        hr = m_EnumIDList->Skip(count);
 
-    if (FAILED_UNEXPECTEDLY(hr))
+        if (FAILED_UNEXPECTEDLY(hr))
+            return hr;
+
+        CComHeapPtr<ITEMIDLIST> spPidl;
+        hr = m_EnumIDList->Next(1, &spPidl, 0);
+        if (FAILED_UNEXPECTEDLY(hr))
+            return hr;
+        hr = ShellObjectCreatorInit<CFolderItem>(m_Folder, spPidl.Detach(), IID_PPV_ARG(FolderItem, ppid));
+        if (FAILED_UNEXPECTEDLY(hr))
+            return hr;
         return hr;
+    }
+    else if (V_VT(&index) == VT_BSTR)
+    {
+        if (!V_BSTR(&index))
+            return S_FALSE;
 
-    CComHeapPtr<ITEMIDLIST> spPidl;
-    hr = m_EnumIDList->Next(1, &spPidl, 0);
-    if (hr == S_OK)
+        HRESULT hr = m_Folder->ParseName(V_BSTR(&index), ppid);
+        if (FAILED_UNEXPECTEDLY(hr))
+            return hr;
+        return hr;
+    }
+    else
     {
-        CFolderItem* item = new CComObject<CFolderItem>();
-        item->AddRef();
-        item->Init(spPidl.Detach());
-        *ppid = item;
-        return S_OK;
+        FIXME("Index type %d not handled.\n", V_VT(&index));
+        return E_NOTIMPL;
     }
-
-    return hr;
+    __debugbreak();
```

Comment:
Does this make some compiler happy?
(you could also just move the "else" code out here to avoid this awkward "unreachable end of non-void function" scenario)

Reply:
It made me happy while debugging, forgot to remove it, thanks :)